File: src/services/instructions/incdec/INC_W_MREG_XX.java
package emulator.source.incdec

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class INC_W_MREG_XX extends Instruction {
	public INC_W_MREG_XX(memory, address, source, 
 destination) {
		super(memory, address, source, destination)
		this.setArgument32()
		super.setAssembler("inc.w [" + this.sdest + " + 0x%08x]")
	}

	
	exec ({ context, memory }) {
		int old = getMemContent(context, fix(context.getReg(this.destination)  + this.argument) / 2, fix(context.getReg(this.destination)  + this.argument))
		long res = old + 1
		setMemContent(context, fix(context.getReg(this.destination)  + this.argument) / 2, (int)res, fix(context.getReg(this.destination)  + this.argument))
		markFlags(res, (int)res, context)
		markOverflow(old, 1, (int)res, context)

		context.pc  += 6
		updateViewer32(context, fix(context.getReg(this.destination)  + this.argument), (int)res)
	}
}

File: src/services/instructions/incdec/IncWMRegXX.js

 import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination) => `inc.w [ ${sdestination} + %s]`

export default class IncWMRegXX extends Instruction {
	constructor (memory, address, source, destination, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		this.setArgument32(memory)
		super.setAssembler(ASSEMBLER_INSTRUCTION_EXPRESSION(this.sdestination), symbolTable)
	}

	exec ({ context, memory }) {
		const old = Instruction.getMemContent(
			context,
			Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument) / 2,
			Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument),
			memory
		)

		const result = old + 1

		Instruction.setMemContent(
			context,
			Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument) / 2,
			result,
			Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument)
		)

		Instruction.markFlags(result, result, context)
		Instruction.markOverflow(old, 1, result, context)

		context.pc  += 6

		// TODO - updateViewer32
		return {
			address: Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  + this.argument),
			content: result
		}
	}
}
